REGRESSION(302506@main): Triggered assertions on visionOS Debug tester

aestes opened the pull request:
#### 49a8cc8d9856ae2e361eef1c05862c27a0c9ca96
<pre>
REGRESSION(302506@main): Triggered assertions on visionOS Debug tester
<a href="https://bugs.webkit.org/show_bug.cgi?id=302013">https://bugs.webkit.org/show_bug.cgi?id=302013</a>
<a href="https://rdar.apple.com/164096712">rdar://164096712</a>

Reviewed by Ryosuke Niwa.

As of 302223@main, subclasses of CanMakeCheckedPtr must be heap-allocated. If they are not (and are
not exempted), CanMakeCheckedPtrBase::~CanMakeCheckedPtrBase crashes due to an assertion failure.
Resolved this by heap-allocating FullScreenWindowControllerVideoPresentationModelClient and storing
it in a unique_ptr in WKFullScreenWindowController.

* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
(-[WKFullScreenWindowController initWithWebView:]):
(-[WKFullScreenWindowController _isBestVideoInFullScreen]):
(-[WKFullScreenWindowController bestVideoFullscreenModeChanged]):
(-[WKFullScreenWindowController fullScreenViewControllerDidInvalidate:]):
(-[WKFullScreenWindowController fullScreenViewController:bestVideoPresentationInterfaceDidChange:]):

Canonical link: <a href="https://commits.webkit.org/302631@main">https://commits.webkit.org/302631@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/895189effcdd78f49aad5dc8bf612495b5a19d10

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/129722 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/1983 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/40578 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/137111 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/81186 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/5c53505c-c3e8-4ec3-bbb7-f0cfb3797885) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/1932 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/1872 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/98820 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/66643 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/6af20e10-acf6-4bc0-a355-8749839c2a65) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/132669 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/1473 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/116180 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/79498 "Passed tests") | | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/96e966c7-94e3-4384-860e-33399c2b0f51) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/1390 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/80383 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/109868 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/34816 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/139593 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/1777 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/1663 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/107327 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/1822 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/112530 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/107201 "Found 1 new API test failure: TestWTF:WTF.DragonBox (failure)") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27292 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/1433 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/31025 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/54528 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/1850 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/65219 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/1664 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/1699 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/1771 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
```diff
@@ -857,7 +857,7 @@ @implementation WKFullScreenWindowController {
 
 #if PLATFORM(VISION)
-    WebKit::FullScreenWindowControllerVideoPresentationModelClient _bestVideoPresentationModelClient;
+    std::unique_ptr<WebKit::FullScreenWindowControllerVideoPresentationModelClient> _bestVideoPresentationModelClient;
```
We probably want to make this `const` and use lazyInitialize to set it since this value is set once and never mutated thereafter.

aestes committed:
REGRESSION(302506@main): Triggered assertions on visionOS Debug tester
https://bugs.webkit.org/show_bug.cgi?id=302013
rdar://164096712

Reviewed by Ryosuke Niwa.

As of 302223@main, subclasses of CanMakeCheckedPtr must be heap-allocated. If they are not (and are
not exempted), CanMakeCheckedPtrBase::~CanMakeCheckedPtrBase crashes due to an assertion failure.
Resolved this by heap-allocating FullScreenWindowControllerVideoPresentationModelClient and storing
it in a unique_ptr in WKFullScreenWindowController.

* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
(-[WKFullScreenWindowController initWithWebView:]):
(-[WKFullScreenWindowController _isBestVideoInFullScreen]):
(-[WKFullScreenWindowController bestVideoFullscreenModeChanged]):
(-[WKFullScreenWindowController fullScreenViewControllerDidInvalidate:]):
(-[WKFullScreenWindowController fullScreenViewController:bestVideoPresentationInterfaceDidChange:]):

Canonical link: https://commits.webkit.org/302631@main
